src/ncp/ncp_base.cpp

```diff
@@ -807,14 +828,175 @@ otError NcpBase::SendQueuedResponses(void)
             // the queue.
 
             mResponseQueueHead = 0;
-            mResponseQueueTail = GetWrappedResponseQueueIndex(mResponseQueueTail);
+            mResponseQueueTail = GetWrappedQueueIndex(mResponseQueueTail, kResponseQueueSize);
         }
     }
 
 exit:
     return error;
 }
 
+// ----------------------------------------------------------------------------
+// MARK: Spinel Pending Command Queue
+// ----------------------------------------------------------------------------
+
+#if OPENTHREAD_CONFIG_MULTIPAN_RCP_ENABLE
+#if OPENTHREAD_RADIO || OPENTHREAD_CONFIG_LINK_RAW_ENABLE
+
+otError NcpBase::EnqueuePendingCommand(PendingCommandType aType, uint8_t aHeader, uint8_t aScanChannel)
+{
+    otError              error = OT_ERROR_NONE;
+    PendingCommandEntry *entry;
+
+    if ((mPendingCommandQueueTail - mPendingCommandQueueHead) >= kPendingCommandQueueSize)
+    {
+        // If there is no room a for a response, emit an unsolicited
+        // `DROPPED` error status to indicate a spinel response was
+        // dropped.
+        mChangedPropsSet.AddLastStatus(SPINEL_STATUS_DROPPED);
+        ExitNow(error = OT_ERROR_NO_BUFS);
+    }
+
+    // Add the new entry in the queue at tail.
+    entry = &mPendingCommandQueue[GetWrappedQueueIndex(mPendingCommandQueueTail, kPendingCommandQueueSize)];
+
+    entry->mType = aType;
+    switch (aType)
+    {
+    case kPendingCommandTypeTransmit:
+        entry->mIid                 = SPINEL_HEADER_GET_IID(aHeader);
+        entry->mTid                 = SPINEL_HEADER_GET_TID(aHeader);
+        entry->mTransmitFrame.mPsdu = entry->mTransmitPsdu;
+        SuccessOrExit(error = DecodeStreamRawTxRequest(entry->mTransmitFrame));
+        break;
+
+    case kPendingCommandTypeEnergyScan:
+        // We dont have access to header, but mCurCommandIID
+        // is updated in HandleReceive, use that instead.
+        entry->mIid         = mCurCommandIID;
+        entry->mScanChannel = aScanChannel;
+        break;
+
+    default:
+        ExitNow(error = OT_ERROR_NOT_FOUND);
+    }
+
+    mPendingCommandQueueTail++;
+
+exit:
+    return error;
+}
+
+otError NcpBase::LoadPendingEntryToTxFrame(const PendingCommandEntry *aEntry)
+{
+    otError       error = OT_ERROR_NONE;
+    otRadioFrame *frame;
+
+    frame = otLinkRawGetTransmitBuffer(mInstance);
+    VerifyOrExit(frame != nullptr, error = OT_ERROR_NO_BUFS);
+
+    VerifyOrExit(aEntry->mTransmitFrame.mLength <= OT_RADIO_FRAME_MAX_SIZE, error = OT_ERROR_PARSE);
+    frame->mLength = aEntry->mTransmitFrame.mLength;
+    memcpy(frame->mPsdu, aEntry->mTransmitFrame.mPsdu, frame->mLength);
+
+    frame->mChannel   = aEntry->mTransmitFrame.mChannel;
```

Comment:
How about we save `frame->mPsdu` pointer in a local variable and then do `*frame = aEntry-> mTransmitFrame` to copy all the `otRadioFrame` fields and then set back the saved `mPsdcu` back on `frame`?

This way we don't need to copy all of the fields one by one (and it will work with any new fields added in future)

Reply:
Make sense. Please let me know if there are further suggestions; thank you.